packages/core/src/common/service-provider.ts

```diff
@@ -0,0 +1,128 @@
+import { collectRecursive, getAllNamedOptional, getOptional } from './inversify-utils';
+import { Proxyable } from './proxy';
+import { serviceIdentifier } from './types';
+
+/**
+ * Represents the `serviceId` string referencing a given service type `T`.
+ */
+export type ServicePath<T> = string & { __staticType: T };
+export function servicePath<T>(path: string): ServicePath<T> {
+    // @ts-expect-error
+    return path;
+}
+
+/**
+ * Part of Theia's Service Layer.
+ *
+ * Whenever a remote wants to use a service over RPC, a request will go through the `ServiceProvider` to find the instance to proxy.
+ */
+export const ServiceProvider = serviceIdentifier<ServiceProvider>('ServiceProvider');
+export interface ServiceProvider {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    getService<T extends object>(serviceId: string | ServicePath<T>, params?: any): T | undefined;
+}
+
+export function bindServiceProvider(bind: interfaces.Bind, contributionName: string | number | symbol): void {
+    bind(ServiceProvider)
+        .toDynamicValue(ctx => {
+            const contributionFilter = getOptional(ctx.container, ContributionFilterRegistry);
+            const contributions = collectRecursive(ctx.container, container => getAllNamedOptional(container, ServiceContribution, contributionName));
+            return new DefaultServiceProvider(contributionFilter?.applyFilters(contributions, ServiceContribution) ?? contributions);
+        })
+        .inSingletonScope()
+        .whenTargetNamed(contributionName);
+}
+
+/**
+ * Part of Theia's Service Layer.
+ *
+ * Requested services to offer over RPC are fetched through a `ServiceProvider` that will source from `ServiceContribution` bindings.
+ *
+ * ## Usage Examples
+ *
+ * ### Record
+ *
+ * ```ts
+ * bind(ServiceContribution)
+ *     .toDynamicValue(ctx => ({
+ *         [PATH1]: () => ctx.container.get(Service1),
+ *         [PATH2]: () => ctx.container.get(Service2),
+ *         [PATH3]: params => ctx.container.get(params.yourParam ? Service3 : Service4);
+ *         // ...
+ *     }))
+ *     .inSingletonScope()
+ *     .whenTargetNamed(YourServiceNamespace);
+ * ```
+ *
+ * ### Function
+ *
+ * ```ts
+ * bind(ServiceContribution)
+ *     .toDynamicValue(ctx => (serviceId, params) => {
+ *         // process arguments...
+ *         return resolvedServiceOrUndefined;
+ *     }))
+ *     .inSingletonScope()
+ *     .whenTargetNamed(YourServiceNamespace);
+ * ```
+ */
+export const ServiceContribution = serviceIdentifier<ServiceContribution>('ServiceContribution');
+export type ServiceContribution = ServiceContributionFunction | ServiceContributionRecord;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type ServiceContributionFunction = (serviceId: string, params?: any) => any;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export interface ServiceContributionRecord { [serviceId: string]: (params?: any) => any };
```

Comment:
Service contributions are one of the new layers: When you want to do RPC rather than explicitly binding services to work over one protocol or the other, you can use this contribution point to provide the instance to expose via RPC, and connection handlers may choose any RPC implementation that will fetch the instance to serve from the contributed services.

e.g. Frontend requests for a proxy at `/jsonrpc/some/service/path` via socket.io. Socket.io receives the new connection and routes it to the `/jsonrpc/*serviceId` connection handler which runs JSON-RPC over the incoming connection. This JSON-RPC connection handler then looks up the service for `serviceId` through `ServiceContribution` bindings, if it find an instance it will wire it to answer JSON-RPC requests over the incoming connection. Done.